tools/inference_engine/src/impl/model.cpp

```diff
@@ -178,6 +187,90 @@ ModelDescriptor::~ModelDescriptor()
     std::cout << "D-TOR ~ModelDescriptor()" << std::endl;
 }
 
+class FusionVisitor : public GpuVisitor 
+{
+    std::unordered_set<GpuNode*> to_delete_;
+    std::unordered_map<GpuNode*, size_t> node_to_index_;
+
+    void mark_node_for_deletion(GpuNode* node) 
+    {
+        to_delete_.insert(node);
+    }
+    void update_indices(std::vector<std::unique_ptr<GpuNode>>& sorted_nodes) 
+    {
+        for (size_t i = 0; i < sorted_nodes.size(); ++i)
+            node_to_index_[sorted_nodes[i].get()] = i;
+    }
+public:
+    void process_sorted_nodes(std::vector<std::unique_ptr<GpuNode>>& sorted_nodes) override 
+    {
+        update_indices(sorted_nodes);
+        for (auto it = std::begin(sorted_nodes); it != std::end(sorted_nodes);) 
+        {
+            to_delete_.clear();
+            (*it)->accept(this);
```

Comment:
can we get rid of this accept(..) pattern and just check do sth like this:
`if dynamic_cast<activation*> { do_activation_stuff(...)`
etc.